edb/server/net_worker.py

```diff
@@ -0,0 +1,243 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2024-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import annotations
+from dataclasses import dataclass
+from enum import Enum
+
+import json
+import typing
+
+import asyncio
+import logging
+
+import httpx
+
+from edb.ir import statypes
+from edb.server.protocol import execute
+from . import dbview
+
+if typing.TYPE_CHECKING:
+    from edb.server import server as edbserver
+    from edb.server import tenant as edbtenant
+
+
+logger = logging.getLogger("edb.server")
+POLLING_INTERVAL = statypes.Duration(microseconds=10 * 1_000_000)  # 10 seconds
+
+
+async def _http(tenant: edbtenant.Tenant) -> None:
+    try:
+        async with asyncio.TaskGroup() as g:
+            for db in tenant.iter_dbs():
+                json_bytes = await execute.parse_execute_json(
+                    db,
+                    """
+                    with
+                        PENDING_REQUESTS := (
+                            select std::net::http::ScheduledRequest
+                            filter .status = std::net::RequestState.Pending
+                        ),
+                        UPDATED := (
+                            update PENDING_REQUESTS
+                            set {
+                                status := std::net::RequestState.InProgress,
+                            }
+                        ),
+                    select UPDATED {
+                        id,
+                        method,
+                        url,
+                        body,
+                        headers,
+                    }
+                    """,
+                    cached_globally=True,
+                )
+                pending_requests: list[dict] = json.loads(json_bytes)
+                for pending_request in pending_requests:
+                    request = ScheduledRequest.from_db_record(pending_request)
+                    g.create_task(handle_request(db, request))
```

Comment:
We should limit the maximum number of concurrent requests here, with probably a config like `net::Config.max_concurrent_net_requests` or something like that (Cloud users shouldn't be able to modify this). Implementation-wise, we can use either an `asyncio.Semaphore` or `asyncio.Queue`, both have some usage in our codebase I think.

Reply:
I'll start out with a hard-coded max in this initial implementation.

Reply:
+1 to what Fantix said.

Reply:
I'm not 100% sure I understand how tenants and instances interact, but would the semaphore be global to this module, or would it want to be scoped within the `_http` function such that it was per-tenant?

Reply:
I think the per-tenant `_http()`-scoped Semaphore is better, as it mitigates the noisy neighbor issue on multi-tenant hosts (one tenant could still issue requests if another tenant is overloading its Semaphore). It does make it slightly harder to estimate system resources allocation when configuring multi-tenant hosts, like you'll need to multiply the per-tenant max connections by the number of tenants, but that's all.

This has a precondition that `_http()` tasks don't overlap for a single tenant (like it is now). Maybe we'll change how this is done with probably a Queue in the future when we add the notification stuff, but I guess it's okay for now.